src/main/java/org/jenkinsci/plugins/kubernetes/credentials/OpenShiftBearerTokenCredentialImpl.java

```diff
@@ -5,170 +5,151 @@
 
-import javax.net.ssl.HostnameVerifier;
-import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.nio.charset.StandardCharsets;
-import java.security.KeyManagementException;
-import java.security.KeyStore;
-import java.security.KeyStoreException;
-import java.security.NoSuchAlgorithmException;
-import java.security.cert.CertificateException;
-import java.security.cert.CertificateFactory;
-import java.security.cert.X509Certificate;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Logger;
 
 /**
  * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
+ * @author Max Laverse
+ * <p>
+ * For the specification, see:
+ * https://docs.openshift.com/enterprise/3.0/architecture/additional_concepts/authentication.html#oauth
  */
 public class OpenShiftBearerTokenCredentialImpl extends UsernamePasswordCredentialsImpl implements TokenProducer {
-
+    // Used to artificially reduce the lifespan of a token
+    protected static final long EARLY_EXPIRE_DELAY = 100;
     private static final long serialVersionUID = 6031616605797622926L;
-
+    private static final Logger logger = Logger.getLogger(OpenShiftBearerTokenCredentialImpl.class.getName());
     private transient AtomicReference<Token> token = new AtomicReference<Token>();
 
     @DataBoundConstructor
     public OpenShiftBearerTokenCredentialImpl(CredentialsScope scope, String id, String description, String username, String password) {
         super(scope, id, description, username, password);
     }
 
-    private Object readResolve() {
-        token = new AtomicReference<Token>();
-        return this;
-    }
-
+    /*
+     * Extract a token from url parameters
+     */
+    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings(value="SF_SWITCH_NO_DEFAULT", justification="Other values can be discarded")
+    protected static Token extractTokenFromLocation(String location) throws TokenResponseError {
+        String parameters = location.substring(location.indexOf('#') + 1);
+        List<NameValuePair> pairs = URLEncodedUtils.parse(parameters, StandardCharsets.UTF_8);
 
-    @Override
-    public String getToken(String serviceAddress, String caCertData, boolean skipTlsVerify) throws IOException {
-        Token t = this.token.get();
-        if (t == null || System.currentTimeMillis() > t.expire) {
-            t = refreshToken(serviceAddress, caCertData, skipTlsVerify);
+        String error = "", errorDescription = "";
+        Token token = new Token();
+        for (NameValuePair pair : pairs) {
+            switch (pair.getName()) {
+                case "access_token":
+                    token.value = pair.getValue();
+                    break;
+                case "expires_in":
+                    try {
+                        token.expire = System.currentTimeMillis() + Long.parseLong(pair.getValue()) * 1000 - EARLY_EXPIRE_DELAY;
```

Comment:
is 100ms enough here - what if I wanted to grab this for use in a build in a pipeline - there could be a few minutes delay before starting a script that needs the credential and the use of the credential?

Reply:
You're right. I overlooked it thought it was seconds. I don't know what a good value would be but 5 minutes sound reasonnable.